src/job_runner/job_runner.cc

```diff
@@ -0,0 +1,146 @@
+#include "cris/core/job_runner/job_runner.h"
+
+#include "cris/core/logging.h"
+#include "cris/core/timer/timer.h"
+
+#include <cstddef>
+#include <cstdlib>
+#include <functional>
+#include <memory>
+#include <thread>
+#include <utility>
+
+namespace cris::core {
+
+thread_local std::atomic<std::uintptr_t> JobRunner::kCurrentThreadJobRunner   = 0;
+thread_local std::atomic<std::size_t>    JobRunner::kCurrentThreadWorkerIndex = 0;
+
+JobRunner::JobRunner(JobRunner::Config config)
+    : config_(std::move(config))
+    , random_engine_(random_device_())
+    , random_worker_selector_(0, config_.thread_num_ - 1) {
+    workers_.reserve(config_.thread_num_);
+    for (std::size_t idx = 0; idx < config_.thread_num_; ++idx) {
+        workers_.push_back(std::make_unique<Worker>(this, idx));
+    }
+    for (auto&& worker : workers_) {
+        worker->Start();
+    }
+}
+
+JobRunner::~JobRunner() {
+    for (auto&& worker : workers_) {
+        worker->Stop();
+        worker->Join();
+    }
```

Comment:
At this point, all worker threads should be stopped, but the ubsan test reported an address misalignment error in the `workers_` destructor. That really seems like a race condition to me. Adding a `CHECK_EQ(ActiveThreadNum(), 0u);` after the for loop can fix the error. I guess it may enforce some ordering here. Do you have any ideas? @xkszltl

Reply:
More details: https://github.com/cyfitech/cris-core/runs/6603923816?check_suite_focus=true#step:10:801

Reply:
Updates: the address somehow drifted in the dtor. Here are my local result (I printed the worker addresses in ctor)

```
E20220526 21:59:53.059895    12 job_runner.cc:25] JobRunner: 0x603000000a30
E20220526 21:59:53.063905    12 job_runner.cc:25] JobRunner: 0x603000000a38
E20220526 21:59:53.067656    12 job_runner.cc:25] JobRunner: 0x603000000a40
E20220526 21:59:53.071564    12 job_runner.cc:25] JobRunner: 0x603000000a48
/usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_iterator.h:991:16: runtime error: reference binding to misaligned address 0x603000000a3c for type 'std::unique_ptr<cris::core::JobRunner::Worker>', which requires 8 byte alignment
0x603000000a3c: note: pointer points here
  30 01 00 00 f0 60 00 00  20 02 00 00 f0 60 00 00  10 03 00 00 f0 60 00 00  03 02 00 00 20 00 00 00
              ^ 
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_iterator.h:991:16 in
```
We are accessing somewhere between the 2nd and 3rd worker pointer... don't understand what is going on...

Reply:
Update: 

Any of these will suppress the problem:

- Adding `CHECK_EQ(ActiveThreadNum(), 0u);`
- Printing addresses of the unique_ptr<Worker> instances
- Printing values of the unique_ptr<Worker> (the pointer to Workers)
- Printing ".joinable()" of the worker threads (if `join()` was called, `joinable()` will become false)
- `CHECK(!worker->thread_.joinable())`

These will NOT suppress it:
- if worker threads joinable then abort
- if ActiveThreadNum() >0 then abort
- get active thread num but not use it (I guess it was optimized out by the compiler)

Reply:
Update: 

Any of these will suppress the problem:

- Adding `CHECK_EQ(ActiveThreadNum(), 0u);`
- Printing addresses of the unique_ptr<Worker> instances
- Printing values of the unique_ptr<Worker> (the pointer to Workers)
- Printing ".joinable()" of the worker threads (if `join()` was called, `joinable()` will become false)
- `CHECK(!worker->thread_.joinable())`

These will NOT suppress it:
- if worker threads joinable then abort
- if ActiveThreadNum() >0 then abort
- get active thread num but not use it (I guess it was optimized out by the compiler)

Reply:
Guess: I looked at the `consume` function of boost lockfree queue, all the atomic operations in it are in "acquire" mode. So we have to make the affected dtor operations to be treated as "affected" ops otherwise they may be reordered before `consume_all`, hence we consume on some invalid addresses.

Reply:
Update: @xkszltl Switching from `std::vector` to `std::array` does not help. Tsan helped to catch another race condition (RNG), but it provides 3 more warnings on boost lockfree queue operations, and failed the following checks:

```
ests/job_runner_test.cc:43: Failure
Expected equality of these values:
  call_count.load()
    Which is: 613
  kJobNum
    Which is: 500
```

How is that possible? `call_count` will never be larger than kJobNum.
```
tests/job_runner_test.cc:45: Failure
Expected equality of these values:
  runner.ActiveThreadNum()
    Which is: 4
  0
```
All workers are still active... I really don't know what that means and what is the execution order. @xkszltl Don't you think this is an either lockfree queue or tsan bug?

Reply:
会不会是boost lock free queue的问题，或者是使用上的问题？换成之前实现的lock queue试试有没有同样的问题出现

Reply:
lock当然没问题，但是性能差很多，lockfree是在刀尖上跳舞，把所有东西都准确串好了才能保证性能和正确性。至于boost的问题，是有可能的，不过是我们自己的错的可能更大。

Reply:
Update:  I printed the worker vector address since the ubsan is complaining about address misalignment of the vector, then a very interesting thing happens:
```
Running main() from gmock_main.cc
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from JobRunnerTest
[ RUN      ] JobRunnerTest.Basic
[New Thread 0x7ffff76c8700 (LWP 370221)]
WARNING: Logging before InitGoogleLogging() is written to STDERR

// These are worker addresses, nothing interesting

E20220531 23:20:45.882959 370204 job_runner.cc:59] JobRunner: 0x3617a0
[New Thread 0x7ffff6ec7700 (LWP 370222)]
E20220531 23:20:45.885442 370204 job_runner.cc:59] JobRunner: 0x4e1ac0
[New Thread 0x7ffff66c6700 (LWP 370223)]
E20220531 23:20:45.887601 370204 job_runner.cc:59] JobRunner: 0x661be0
[New Thread 0x7ffff5ec5700 (LWP 370224)]
E20220531 23:20:45.889887 370204 job_runner.cc:59] JobRunner: 0x7e1f00

// Here is &workers_, looks normal when construction

E20220531 23:20:45.889925 370204 job_runner.cc:61] JobRunner: 0x7fffffffdd20
[Thread 0x7ffff76c8700 (LWP 370221) exited]
[Thread 0x7ffff6ec7700 (LWP 370222) exited]
[Thread 0x7ffff66c6700 (LWP 370223) exited]
[Thread 0x7ffff5ec5700 (LWP 370224) exited]

// All other workers threads have exited. The main thread is the only active one.

Thread 1 "job_runner_test" hit Breakpoint 1, cris::core::JobRunner::~JobRunner (this=<optimized out>) at src/job_runner/job_runner.cc:90
90      }
(gdb) p workers_
value has been optimized out
(gdb) f 1
#1  0x000000000022d1d0 in cris::core::JobRunnerTest_Basic_Test::TestBody (this=<optimized out>) at tests/job_runner_test.cc:86
86      }

// The workers_ address looks correct even at this time

(gdb) p &runner.workers_
$1 = (cris::core::JobRunner::worker_list_t *) 0x7fffffffdd20

// One step further, we head into the dtor of the workers_ vector, now the address looks weird... It drifted 1 bit to 0x7fffffffdd21, that is why it is misaligned.

(gdb) s
std::vector<std::unique_ptr<cris::core::JobRunnerWorker, std::default_delete<cris::core::JobRunnerWorker> >, std::allocator<std::unique_ptr<cris::core::JobRunnerWorker, std::default_delete<cris::core::JobRunnerWorker> > > >::~vector (this=0x7fffffffdd21)
    at /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_vector.h:679
679           {

// thread 1 is the only active thread.

(gdb) info threads
  Id   Target Id                                            Frame 
* 1    Thread 0x7ffff76d7240 (LWP 370204) "job_runner_test" std::vector<std::unique_ptr<cris::core::JobRunnerWorker, std::default_delete<cris::core::JobRunnerWorker> >, std::allocator<std::unique_ptr<cris::core::JobRunnerWorker, std::default_delete<cris::core::JobRunnerWorker> > > >::~vector (this=0x7fffffffdd21) at /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/stl_vector.h:679

// Now the vector address is drifted, but if you switch to the upper frame and check the address from there, it is still normal!!

(gdb) f 1
#1  0x000000000022d1d0 in cris::core::JobRunnerTest_Basic_Test::TestBody (this=<optimized out>) at tests/job_runner_test.cc:86
86      }
(gdb) p &runner.workers_                                                                                                                                                                                                                                                               
$2 = (cris::core::JobRunner::worker_list_t *) 0x7fffffffdd20
```

Reply:
Tested on gcc's ubsan and I don't see any issue again. Maybe there are bugs in the llvm's ubsan or we have to link it with libc++? Any ideas? @xkszltl

Reply:
Update: `asan` failure is triggered by missing copt and c++runtime linking. `ubsan` issue may also be because of my local change on linkopts. The ubsan failure at the beginning was fixed by the locks and fences.

Reply:
The magic bit is `EVENTUALLY_EQ`...

Reply:
call count is out of lifecycle...